src/Internal/DependencyManagement/ExtendedContainer.php

```diff
@@ -0,0 +1,102 @@
+<?php
+/**
+ * ExtendedContainer class file.
+ *
+ * @package Automattic\WooCommerce\Internal\DependencyManagement
+ */
+
+namespace Automattic\WooCommerce\Internal\DependencyManagement;
+
+use Automattic\WooCommerce\Container;
+use League\Container\Definition\DefinitionInterface;
+
+/**
+ * This class extends the original League's Container object by adding some functionality
+ * that we need for WooCommerce.
+ */
+class ExtendedContainer extends \League\Container\Container {
+
+	/**
+	 * Whitelist of classes that we can register using the container
+	 * despite not belonging to the WooCommerce root namespace.
+	 *
+	 * @var string[]
+	 */
+	private $registration_whitelist = array(
+		\Psr\Container\ContainerInterface::class,
+	);
+
+	/**
+	 * Register a class in the container.
+	 *
+	 * @param string    $id Class name.
+	 * @param mixed     $concrete How to resolve the class with `get`: a factory callback, a concrete instance, onother class name, or null to just create an instance of the class.
+	 * @param bool|null $shared Whether the resolution should be performed only once and cached.
+	 *
+	 * @return DefinitionInterface The generated definition for the container.
+	 * @throws \Exception Invalid parameters.
+	 */
+	public function add( string $id, $concrete = null, bool $shared = null ) : DefinitionInterface {
```

Comment:
Is `$id` necesarrily name of the class? For egs, can we have multiple shared instances of same class identified by different IDs. If not, then we should rename the param from `$id` to `$class_name` everywhere. If yes then we need to remove this check.